### ISSUE TITLE
group output error

### ISSUE THREAD

**Issue by kbilsted** [houmain/keymapper#138]
hi

there are problems when using two group definitions

i have remapped **all my keys such that I am no longer using qwerty** and as the first i define 

    (N I O ) >> F O U N D   
    (A S D) >> I O N  

when *both* definitions are in use it triggers the error.. 

the error happens when pressing eg `s d` fast which produce `id`..  

without the above defs it produces `in` (which is correct according to other definitions).

**Comment by ristomatti** [comment]
I'll try if I can reproduce it. So the full config to would be this?


```ini
I >> U
O >> O
S >> I
D >> N

(N I O ) >> F O U N D   
(A S D) >> I O N 
```

**Comment by ristomatti** [comment]
At least with that, I can't get reproduce the output of `id`:

```
inininininininininininniniininin
inininininininininininniniininin
ininininininininin
```

Out of curiosity I tried the `(A S D)` combination:

```
ionionionnioinonioion
ionionionnioinonioion
```

And the `(N I O)`:

```     
foundofundfounduofnd
               ^^^^^
foundfoundfounduofndfounduofnduofndufond
               ^^^^^     ^^^^^^^^^^^^^^^
```

I then tried it again with some more thought. I can consistently make `nio` output `found` if I press the keys down in that order, e.g `N{I{O}}` but if I press them the other way round `O{I{N}}` the output ends up being `oufnd` every time. I get another "mispelling" if I try to press `I` down first.

`asd` behaves the same way. The defined output comes in the correct order only if the keys get pressed like `A{S{D}}`. Intresting! I can only imagine what you're trying to make out of this or is it just testing if Keymapper can do it or not? :smile:

**Comment by ristomatti** [comment]
I tried changing the last lines to this:

```
(N I O ) >> 'found'   
(A S D) >> 'ion'
```

...but still get the same behavior.

**Comment by kbilsted** [comment]
thanks for testing. no all keys are remmapped..  i found the bug because i kept "misspelling" and it made no sense

**Comment by kbilsted** [comment]
try 

```
(N I O ) >> F O U N D   
(A S D) >> I O N  
I >> U
A >> S 
S >> I
D >> N
N >> F
```

also found out if i press and hold `d` and press `s` (rolling the oher way) i get `ns`

**Comment by kbilsted** [comment]
version 4.1.1

**Comment by ristomatti** [comment]
The combined keys seem to work pretty much flawlessly when defined first. But I can reproduce what you described both ways.

- hold `S` then press `D`, output is `in`
- hold `D` then press `S`, ouput is `ni`
- roll from `S` to `D`, ouput is `id`
- roll from `D` to `S`, ouput is `ns`

**Comment by ristomatti** [comment]
I believe getting this right in all use cases might be really difficult.

In addition to that issue, there's the compromise with the key combinations. If defined afterwards, the output can get mixed up and if defined first, you lose key repeat of any of the keys used in the combination (which to be fare, is to be expected).

**Comment by ristomatti** [comment]
I find the idea of using Keymapper to define an alternative layout somewhat difficult to grasp/maintain on the config side. If I think about it intuitively, you'd first define to what key each singular key maps to. Then when adding further mappings, you would do it based on the mapping defined earlier, but it doesn't work like as each key press in interpreted by the key code, right?

So if I'd do e.g. a dvorak/workman layout for the alphas, then swap the keycaps accordingly, I would still need to do the following mappings using the original key names (= keycodes sent by the keyboard).

Unless I've got this completely wrong, I'd think to make the process more intuitive, the key handling would need to be done in two phases based on two separate config files. First would define the singular key remaps, and the second would work the same way as it works now, with the exception that the mappings could be defined based on remaps done in the first phase.

**Comment by kbilsted** [comment]
@ristomatti Possibly other tools are more suited, but eg kanata has a hard time with using ALT as a layer swap key while at the same time allows the alt-tab, or the alt+mouse = column select so I tried Keymapper and this is where I am at. If you can advice on kmonad or kanata setup ill gladly try remapping using those.

I don't swap any key caps I learn the keyboard by heart and sometime use my always-on-keyboard display ;)

One could do multiple configurations as you suggest, but perhaps we are bending the tool to where it doesn't want to be. In any case my complaints were about unmatched groups having effect on the output, not changing the whole configuration language

**Comment by kbilsted** [comment]
@ristomatti  when the layout has matured more I'll gladly share it with others. It's awesome so far.. I just need to get to 50 WPM before I can judge if it is insanity or ingenious what I'm doing! drop me an email for private further conversations rollercoasterlayout@gmail.com

**Comment by houmain** [comment]
Thanks for helping with the reproduction. This should be fixed in the 4.1.3 release. I mean it should no longer forward keys unmapped when it can no longer match the group.

> the key handling would need to be done in two phases based on two separate config files.

I have also thought about that once and might implement it some day.

**Comment by ristomatti** [comment]
@kbilsted New release just out of the oven, claims to fix this issue https://github.com/houmain/keymapper/releases/tag/4.1.3.

@houmain In case you're not aware of this, GitHub can automatically link commits to issues. This gives instant feedback to people watching an issue.

For example, in the release notes for 4.1.3 you mention this issue has been fixed. By using specific keywords in your commit messages, the commits can be linked to the related issues. In this case, if the commit message of https://github.com/houmain/keymapper/commit/81312c781f592a24e4530696279e6854be4448e2 would have contained the magic words "fixes #138", the commit would be shown on this issue therad. 

There's plenty of alternative phrases to use: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword.

It's a bit unclear to me though if a pull request needs to be made for this to occur if it's enough to just commit on the main branch.

**Comment by ristomatti** [comment]
> I have also thought about that once and might implement it some day.

Somehow I'm not surprised. 😄

I'll test the fix out of curiosity.

@houmain Did you notice https://github.com/houmain/keymapper/discussions/139? It needs to be updated to include `--set-config` though. I'm happy to see `--no-notify` as well. I use Rofication for notications and might make me miss other notifications if I'm eperimenting with a config. But does it also silence the notification of an invalid config?

**Comment by ristomatti** [comment]
I can no longer reproduce either issue when config is defined like this:

```ini
(N I O ) >> F O U N D   
(A S D)  >> I O N  

I >> U
A >> S 
S >> I
D >> N
N >> F
```

If I move the two first lines in the end, the output is not consistent:

- pressing `asd` by rolling the keys in that order results in `sin` 
- pressing `dsa` the same way results in `nis`
- pressing `nio` the same way results in `fuond`
- pressing `oin` the same way results in `found`

I don't know if this is to be expected.

**Comment by kbilsted** [comment]
Lol who knows ;)

**Comment by ristomatti** [comment]
I did a little experiment to attempt fixing key repeat. While doing it, I noticed [a bug](https://github.com/houmain/keymapper/issues/135#issuecomment-2089064404), but also got some interesting results to mess up with our brains.

This is the config:

```ini
I !100ms I >> U U
A !100ms A >> S S 
S !100ms S >> I I
D !100ms D >> N N
N !100ms N >> F F

(N I O) >> F O U N D   
(A S D) >> I O N  

I >> U
A >> S 
S >> I
D >> N
N >> F
```

With this config, the individual key remaps seem to work as expected and so does my key repeat fix attempt. But the grouped keys give a rather interesting outcome 😬:

Pressing `oin` in a rolling manner results in `found` as one would expect, but doing it the other way (`ion`) results in `fuond` (the same output as I got from `nio` described on my previous comment).

With the `(A S D)` things get interesting.

Rolling `asd` outputs `sion` while rolling `dsa` results in `nion`. :thinking:

**Comment by ristomatti** [comment]
I did a further experiment, changing the combined keys in the above config to:

```
N{I{O}} >> F O U N D
A{S{D}} >> I O N
```

This obviously breaks pressing them at the same time or in a reverse order. When typed in the specified order, the output is exactly the same as described above.

When pressed the same time in this sequence: `asd` -> `fuond`. And as previously `nio` -> `sion`. It would be interesting to know how the internal processing ends up with the extra character.

**Comment by ristomatti** [comment]
Given the amount of things occurring at the same time, I would almost be ready to just accept this to be "unfixable" and call it a day. What do you think @houmain, is it just a bug waiting to be fixed or does this involve some OS/hardware level randomness that just appears in a different place when one issue is fixed?

**Comment by houmain** [comment]
> is it just a bug waiting to be fixed or does this involve some OS/hardware level randomness that just appears in a different place when one issue is fixed?

I tried these configurations a bit. The results may not be what the creator intended, but I don't think they are illogical.
For example, with this configuration:
```
I >> U
A >> S 
S >> I
D >> N
N >> F

(N I O ) >> F O U N D   
(A S D)  >> I O N  
```
rolling over `N I O` results in `fuond`, 
rolling over `I N O` results in `ufond`,
rolling over `O I N` and `O N I` result in `found`.

The first two immediately output something as soon as they are pressed (`F` and `U` in the corresponding order).
The last two wait after the `O`, whether more keys of the group follow.
In the first two cases, when the group is completed only `O N D` need to be pressed, since `F` and `U` are already down.

**Comment by houmain** [comment]
> Does it also silence the notification of an invalid config though?

Yes, `--no-notify` silences all notifications.

**Comment by ristomatti** [comment]
Perhaps consider a flag that only silences success notifications? I still enabled the flag though. Most of the times when I edit the mappings, I have my `ContextActive` debug log visible. If I see nothing logged after saving, I know I messed up and I'll need to run `keymapper --check`. 🙂

Normally just an undo would be sufficient, but I've recently changed Helix as my default text editor and it's way of selecting text on movement/search still gets me off guard too often. I then end up unknowingly editing multiple lines at the same time, some of which might be hundreds of lines off screen. :grin:

**Comment by kbilsted** [comment]
@houmain  are you interested in working on supporting "adaptive keys" as describe above ?

**Comment by houmain** [comment]
Yes, I think I could implement something like https://github.com/houmain/keymapper/issues/137#issuecomment-2086300637 and then I could also add a way to access the recently typed keys.

**Comment by ristomatti** [comment]
> The results may not be what the creator intended, but I don't think they are illogical.

Now it does make sense indeed, thanks for the step by step explanation. Luckily I'm not interested in trying to implement anything as exotic (at least for now) as my brain would melt debugging issues like this. :sweat_smile: 

Accessing the runtime value of `Any` or aliased keys would be great for debugging as well if it would mean the pressed key could be used on the output side e.g. to log it within an executed script.

**Comment by kbilsted** [comment]
you may be right.

**Comment by ristomatti** [comment]
@kbilsted Was the reply for me or @houmain and to what comment?

**Comment by houmain** [comment]
A few new features are available in the 3.4.0 release. 

> The results may not be what the creator intended, but I don't think they are illogical.

`!Any` allows to release already pressed keys before outputting the new ones. Which should make the result less surprising:

```
I >> U
A >> S 
S >> I
D >> N
N >> F

(N I O ) >> !Any F O U N D   
(A S D)  >> !Any I O N  
```
rolling over `N` `I` `O` results in `fufound`.

> Yes, I think I could implement something like [#137 (comment)](https://github.com/houmain/keymapper/issues/137#issuecomment-2086300637) 

One can write now:
```
? C A T >> Backspace Backspace D O G
```
The mapping is skipped as long as it does not match completely. So it does not wait after the `C` whether a `A` follows...

**Comment by ristomatti** [comment]
I'll leave the honor of testing this first to you @kbilsted. 😁
